pyteal/ast/opup.py

```diff
@@ -133,13 +139,21 @@ def ensure_budget(self, required_budget: Expr) -> Expr:
         return Seq(
             buffered_budget.store(required_budget + buffer),
             While(buffered_budget.load() > Global.opcode_budget()).Do(
-                self._construct_itxn()
+                self._construct_itxn(inner_fee=inner_fee)
             ),
         )
 
-    def maximize_budget(self, fee: Expr) -> Expr:
+    def maximize_budget(self, fee: Expr, inner_fee: Expr = None) -> Expr:
```

Comment:
I think the change here needs more thought. We already have the user specify `fee`, so `inner_fee` doesn't make sense.

What I think we really want is a way of specifying _who_ should pay the fee. Here are the options I think make sense (notice these are the same as `ensure_budget` when `inner_fee` is None, 0, and min txn fee):
1. Fee is first taken from the group fee credit, if there is one, then drawn from the caller app's account balance if needed
2. Fee is only taken from the group fee credit -- error if the fee credit is too small
3. Fee is only taken from the caller app's account balance

Reply:
I think this is done with the latest push, will add docs if it looks reasonable

Reply:
@barnjamin Checking my understanding:  Is it intentional that `ensure_budget` does _not_ use `OpUpFeePayer`?

I assume `ensure_budget` _should_ reference `OpUpFeePayer`.

Reply:
That probably makes more sense than what it is currently, I was thinking the change to add `OpUpFeePayer` was meant to disambiguate fee/inner_fee of the `maximize_budget` but since I don't expect the inner_fee would be anything but 0/1000/None we should probably use the same logic as maximize